File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    static void findCombination(int idx,int[] candidates,int target,List<List<Integer>> result, List<Integer> combination){
        if(idx == candidates.length || target == 0){
            if(target == 0){
                result.add(new ArrayList<>(combination));
            }
            return;
        }

        for(int i = idx; i < candidates.length; i++){
            if( i > idx && candidates[i-1] == candidates[i]) continue;
            if(candidates[i] > target)    break;

            combination.add(candidates[i]);
            findCombination(i+1, candidates, target-candidates[i], result,combination);
            combination.remove(combination.size() - 1);
        }
        
    }
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(candidates);
        findCombination(0, candidates, target, result, new ArrayList<>());
        return result;
    }
}

File: 0040-combination-sum-ii/0040-combination-sum-ii.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function(candidates, target) {
    candidates.sort((a, b) => a - b);
    var result = [];
    generateCombinationSum(result,[],0,candidates,target);
    return result;
};

function generateCombinationSum(result,current,j,candidates,target){
    if(j == candidates.length || target == 0){
        if(target == 0){
            result.push([...current]);
        }
        return;
    }

    for(let i =j; i<candidates.length; i++){
        if(i != j && candidates[i] == candidates[i-1])  continue;
        if( candidates[i] > target) continue;

        current.push(candidates[i]);
        generateCombinationSum(result,current,i+1,candidates,target-candidates[i]);
        current.pop();
    }
}